### PR TITLE
[NVIDIA XLA] FP8 Matmul BiasAdd fusion for high rank input

### PR DESCRIPTION
**Pull request by wenscarl**
This PR complete the coverage for FP8 matmul when input is high rank tensor. @philipphack

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 1]
```diff
@@ -516,11 +516,12 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
     //
-    if (Match(instr,
-              m::AddAnyOrder(
-                  m::Bitcast(CublasLtMatmul(&existing_gemm).WithOneUser())
-                      .WithOneUser(),
-                  m::Broadcast(&bias, m::Op()).WithOneUser()))) {
+    if (Match(
+            instr,
+            m::AddAnyOrder(
+                m::Bitcast(CublasLtMatmulMaybeF8(&existing_gemm).WithOneUser())
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:522]: This optimization is only valid if the bitcast does not change the shape, and is only valid for certain layouts. I haven't thought through what layouts this optimization is valid for, but it's valid if all layouts have the dimensions ordered from most major to most minor.

CC @philipphack

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 2]
```diff
@@ -567,8 +569,9 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     }
 
     if (Match(instr,
-              m::AddAnyOrder(GemmOrCublasLtMatmul(&existing_gemm).WithOneUser(),
-                             m::Op(&bias).WithPredicate(is_not_broadcast)))) {
+              m::AddAnyOrder(
+                  GemmOrCublasLtMatmulMaybeF8(&existing_gemm).WithOneUser(),
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:573]: I think the logic in `CreateF8CustomCall` that supports matrix bias fusions should be removed, since it's redundant with this logic

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
Hi @wenscarl Can you please resolve conflicts? Thank you!

**Comment by reedwm** [comment]
This PR is not stale. It requires https://github.com/tensorflow/tensorflow/pull/60409 to be merged first. I'll reopen if automatically closed.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 3]
```diff
@@ -1103,7 +1125,52 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
-    operands.insert(operands.begin() + 2, MaybeConstantFoldBias(bias));
+    HloInstruction *broadcast_bias = MaybeConstantFoldBias(bias);
+    if (slice) {
+      broadcast_bias = instr->AddInstruction(
+          HloInstruction::CreateBitcast(slice->shape(), broadcast_bias));
+    }
+    absl::Span<const int64_t> batch_dims =
+        config.dot_dimension_numbers().rhs_batch_dimensions();
+    // Get the padded shape.
+    auto pad_shape = [&batch_dims](const Shape old_shape) {
```
Comment (philipphack) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1136]: Does this duplicate functionality from CreateF8CustomCall? Can the lambda be replaced by a function?

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 4]
```diff
@@ -1103,7 +1125,52 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
+              RoundUpTo<int64_t>(old_shape.dimensions(i), 16);
+          padded_shape.set_dimensions(i, padded_dimension);
+        }
+      }
+      return padded_shape;
+    };
+
+    // Pad the non-batch dimensions of the operands to multiples of 16 as
+    // required by cuBLASLt.
+    auto pad_operand = [&instr, &pad_shape](HloInstruction *&x) -> void {
```
Comment (philipphack) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1150]: Same as above. Also, are we still fusing the matrix bias in CreateF8CustomCall?

**Review by philipphack — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 3 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc]: Placed by functions in newer commit.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 4 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc]: No. All matrix bias fusion have been moved out of CreateF8CustomCall.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 2 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:589]: It's been moved to FuseMatrixBiasAdd.

**Comment by wenscarl** [comment]
@reedwm The top of master seems to fail those tests in `gemm_rewrite_test.cc` which prevents from rebasing.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 1 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:538]: The `bitcast` stands there even for non-fp8. So should fp8 be treated differently here?

Comment (philipphack) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:538]: I think this should be fine if the bitcast acts as a reshape.

Comment (thomasjoerg) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:538]: @reedwm Please get back to wenscarl and Philipp regarding the change you requested here.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 2 above)
Comment (thomasjoerg) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:589]: @reedwm please resolve the conversation or get back to wenscarl.

**Comment by gbaned** [comment]
Hi @reedwm Can you please review this PR ? Thank you!

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 5]
```diff
@@ -1046,8 +1054,11 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
 
   Status FuseMatrixBiasAdd(HloInstruction *instr, HloInstruction *bias,
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1055]: Please comment the structure of instructions this function expects. E.g., the comment can be

```
Fuses a matrix bias into a cuBLAS call. 'instr' should be an Add instruction in the following form:
    Add(OptionalBitcast(OptionalSlice(gemm)), bias)
'gemm' is expected to be a cuBLAS custom_call.
```

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 6]
```diff
@@ -1046,8 +1054,11 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
                            const HloInstruction *gemm,
-                           HloInstruction *bitcast = nullptr) {
-    TF_RET_CHECK(bias->shape() == (bitcast ? bitcast->shape() : gemm->shape()));
+                           HloInstruction *bitcast = nullptr,
+                           HloInstruction *slice = nullptr) {
+    TF_RET_CHECK(
+        (bias->shape() == (bitcast ? bitcast->shape() : gemm->shape())) ||
+        (bias->shape() == (slice ? slice->shape() : gemm->shape())));
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1061]: Only check the slice shape if there is no bitcast, since otherwise this condition is confusing. You can write this as:

```c++
TF_RET_CHECK(bias->shape() == (bitcast ? bitcast->shape() : slice ? slice->shape() : gemm->shape()));
```

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 7]
```diff
@@ -1131,6 +1154,11 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
           ->set_output_to_operand_aliasing({{{}, {2, {}}}});
     }
     TF_RETURN_IF_ERROR(SetName(instr->GetModule(), fused_op.get()));
+    if (slice != nullptr) {
+      fused_op = slice->CloneWithNewOperands(
+          slice->shape(),
+          {slice->parent()->AddInstruction(std::move(fused_op))});
+    }
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1161]: I don't think moving the slice after the Add op is necessarily valid. Adding before the slice gives a different result than adding after the slice.

`FuseVectorBiasAdd` has the same issue today, even without this PR. The following JAX program outputs an incorrect matrix due to this issue

```python
import os
import jax.numpy as jnp
import jax
os.environ['XLA_FLAGS'] = '--xla_gpu_enable_cublaslt=true'


@jax.jit
def f(x, y, b):
  output = jnp.matmul(x, y)
  output_sliced = output[1:, 1:]
  output_biased = output_sliced + b
  return output_biased


x = jnp.array([[1., 1.],
               [1., 1.],
               [1., 1.]])
y = jnp.array([[1., 1., 1.],
               [1., 1., 1.]])
b = jnp.array([1., 2.])
output_biased = f(x, y, b)
print(output_biased)
```

It outputs `[[4, 2], [4, 2]]` when it should output `[[3, 4], [3, 4]]`. Removing `@jax.jit` causes it to output the correct value.

I think restricting the slice to only be at the start of each dimension will fix the issue, although I haven't proved this to myself yet.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 8]
```diff
@@ -1145,10 +1173,12 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
                                    HloInstruction *broadcast,
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1173]: As before, please comment the structure of instructions this function expects. E.g., the comment can be

```
Fuses a vector bias into a cuBLAS call. 'instr' should be an Add instruction in the following form:
    Add(OptionalBitcast(OptionalSlice(gemm)), Broadcast(OptionalConvert(...)))
'gemm' is expected to be a cuBLAS custom_call. 
```

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 9]
```diff
@@ -1173,20 +1203,8 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
-      if (it == broadcast_dims.end()) {
-        return false;
-      }
-
-      int64_t vector_dim = it - broadcast_dims.begin();
-      if (bias->shape().layout().minor_to_major(i) != vector_dim) {
-        return false;
-      }
+    if (bitcast) {
+      broadcast_dims = gemm->shape().dimensions();
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1207]: Why remove the existing logic here?

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 1 above)
Comment (reedwm) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:538]: I couldn't find a case where a bitcast did not act a reshape. I am ok with assuming it acts as a reshape for now.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 10]
```diff
@@ -573,11 +585,26 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
+    HloInstruction *optional_slice_matrix = nullptr;
+    if (Match(instr,
+              m::AddAnyOrder(
+                  OptionalBitcast(
+                      &optional_bitcast_matrix,
+                      OptionalSlice(
+                          &optional_slice_matrix,
+                          CublasLtMatmulMaybeF8(&existing_gemm).WithOneUser()))
+                      .WithOneUser(),
+                  m::Op(&bias).WithPredicate(is_not_broadcast)))) {
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:604]: This Match statement looks like it matches a superset of the previous Match statement. If that is the case, remove the previous match statement.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 11]
```diff
@@ -634,6 +661,43 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     return OkStatus();
   }
 
+  // Get the padded shape.
+  Shape pad_shape(const Shape old_shape,
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:665]: Call this PadShapeToMultipleOf16. And make it a top-level function instead of a method. And document it with more detail, e.g.

"Returns a new shape with non-batch dimensions padded to multiples of 16, as required by cuBLASLt FP8 gemms"

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 12]
```diff
@@ -634,6 +661,43 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
+            RoundUpTo<int64_t>(old_shape.dimensions(i), 16);
+        padded_shape.set_dimensions(i, padded_dimension);
+      }
+    }
+    return padded_shape;
+  }
+
+  // Pad the non-batch dimensions of the operands to multiples of 16 as
+  // required by cuBLASLt.
+  void pad_operand(absl::Span<const int64_t> batch_dims, HloInstruction *&instr,
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:680]: Call this PadOperandToMultipleOf16. And make it a top-level function instead of a method. And have it return the new instruction instead of taking it as a reference argument. And don't pass `instr` by reference. And document this function with more detail, e.g.

"Pad the non-batch dimensions of the operands to multiples of 16, as required by cuBLASLt FP8 gemms."

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 13]
```diff
@@ -1233,6 +1251,10 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
       }
     }
 
+    if (bitcast) {
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1254]: I think this should be `slice`, not `bitcast`.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 13 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc]: When the operations of both `bitcast` and `slice` are present, it indicates the presence of high rank input.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 9 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 7 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1159]: The padding operation exclusively adds zeros at the end, resulting in the slice operation only slicing the index from start. For those cases, the results are correct. So I believe aforementioned scenario is unlikely to occur here.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 5 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1052]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 6 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 8 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1171]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 10 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 11 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 14]
```diff
@@ -1145,10 +1172,19 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
                                    HloInstruction *slice = nullptr,
-                                   HloInstruction *convert = nullptr) {
-    TF_RET_CHECK(ShapeUtil::Compatible(
-        broadcast->shape(), (slice ? slice->shape() : gemm->shape())));
-
+                                   HloInstruction *convert = nullptr,
+                                   HloInstruction *bitcast = nullptr) {
+    // Fuses a vector bias into a cuBLAS call. 'instr' should be an Add
+    // instruction in the following form:
+    // Add(OptionalBitcast(OptionalSlice(gemm)), broadcast), where 'gemm' is
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1179]: The pattern should have the fact the broadcast is of the optional convert. It should be:

```
Add(OptionalBitcast(OptionalSlice(gemm)), Broadcast(OptionalConvert(...)))
```

Also similar to my comment on `FuseMatrixBiasAdd`, it's clearer if the pattern is on its own line.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 15]
```diff
@@ -1173,19 +1209,23 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     // dimensions; i.e. its most minor physical dimensions align with most minor
     // physical dimensions of the gemm output.
     absl::Span<const int64_t> broadcast_dims = broadcast->dimensions();
-    for (size_t i = 0; i < num_col_dims; ++i) {
-      int64_t dim = gemm->shape().layout().minor_to_major(i);
+    if (bitcast != nullptr) {
+      broadcast_dims = gemm->shape().dimensions();
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1213]: Why not check the bias vector is broadcast in the most major dimensions if there is a bitcast?

Also why set `broadcast_dims` in this branch? It is only used in the `else` branch.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 16]
```diff
@@ -1145,10 +1172,19 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
-        broadcast->shape(), (slice ? slice->shape() : gemm->shape())));
-
+                                   HloInstruction *convert = nullptr,
+                                   HloInstruction *bitcast = nullptr) {
+    // Fuses a vector bias into a cuBLAS call. 'instr' should be an Add
+    // instruction in the following form:
+    // Add(OptionalBitcast(OptionalSlice(gemm)), broadcast), where 'gemm' is 
+    // expected to be a cuBLAS custom_call. The optional convert is only 
+    // applicable to F8 matmul as cublasLt has specific constraints on the 
+    // vector bias type. The optional bitcast is necessary to handle high rank
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1182]: I would rephrase this sentence as "optional_convert is only used for F8 matmuls as cuBLASLt has specific constraints on the vector bias type for such matmuls."

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 17]
```diff
@@ -494,39 +538,44 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
-    // including when slicing is applied to the result.
+    HloInstruction *optional_bitcast = nullptr;
+    // Attempt to elide broadcast and fuse addition of a vector bias into
+    // GEMM, including when slicing is applied to the result.
     if (Match(instr,
               m::AddAnyOrder(
-                  OptionalSlice(
-                      &optional_slice,
-                      CublasLtMatmulMaybeF8(&existing_gemm).WithOneUser())
+                  OptionalBitcast(
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:549]: Out of curiosity, do you know why a bitcast is being introduced here?

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 18]
```diff
@@ -1044,10 +1046,17 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     return OkStatus();
   }
 
+  // Fuses a matrix bias into a cuBLAS call. 'instr' should be an Add 
+  // instruction in the following form:
+  // Add(OptionalBitcast(OptionalSlice(gemm)), bias) where 'gemm' is expected
```
[the remark this answers is not in the document]
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1051]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 19]
```diff
@@ -1103,7 +1113,19 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
 
     std::vector<HloInstruction *> operands(gemm->operands().begin(),
                                            gemm->operands().end());
-    operands.insert(operands.begin() + 2, MaybeConstantFoldBias(bias));
+    HloInstruction *broadcast_bias = MaybeConstantFoldBias(bias);
```
[the remark this answers is not in the document]
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1116]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 20]
```diff
@@ -1103,7 +1113,19 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
+    if (slice && bitcast) {
+      broadcast_bias = instr->AddInstruction(
+          HloInstruction::CreateBitcast(slice->shape(), broadcast_bias));
+    }
+
+    if (gemm->custom_call_target() == kCublasLtMatmulF8CallTarget) {
+      broadcast_bias = PadOperandToMultipleOf16(
+          config.dot_dimension_numbers().rhs_batch_dimensions(), instr,
+          broadcast_bias);
+    }
```
[the remark this answers is not in the document]
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1126]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 14 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1179]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 15 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1213]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 16 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1182]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 21]
```diff
@@ -1103,7 +1123,20 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
 
     std::vector<HloInstruction *> operands(gemm->operands().begin(),
                                            gemm->operands().end());
-    operands.insert(operands.begin() + 2, MaybeConstantFoldBias(bias));
+    HloInstruction *maybe_constant_folded_bias = MaybeConstantFoldBias(bias);
+    if (slice && bitcast) {
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1127]: I think we should create a bitcast even if there is no slice. As long as there is a bitcast, we need to bitcast the bias.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 22]
```diff
@@ -1103,7 +1123,20 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
+      maybe_constant_folded_bias =
+          instr->AddInstruction(HloInstruction::CreateBitcast(
+              slice->shape(), maybe_constant_folded_bias));
+    }
+
+    if (gemm->custom_call_target() == kCublasLtMatmulF8CallTarget ||
+        gemm->custom_call_target() == kCublasLtMatmulCallTarget) {
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1134]: Shouldn't we pad for cublas non-LT gemms as well?

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 23]
```diff
@@ -1131,6 +1164,16 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
           ->set_output_to_operand_aliasing({{{}, {2, {}}}});
     }
     TF_RETURN_IF_ERROR(SetName(instr->GetModule(), fused_op.get()));
+    if (slice != nullptr) {
+      int slice_op_dim = slice->operand(0)->shape().rank();
+      if (slice->slice_starts() != std::vector<int64_t>(slice_op_dim, 0)) {
+        return OkStatus();
+      }
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1171]: Check that the strides are 1. Also better to move this check to the start of the function, so you don't unnecessarily create extra ops. Also add a comment stating "Only slices which chop off the ends of dimensions are supported".

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 14 above)
Comment (reedwm) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc]: You still don't show the OptionalConvert in the pattern

File: tensorflow/compiler/xla/service/gpu/tests/gemm_rewrite_test.cc [hunk 24]
```diff
@@ -5323,6 +5323,388 @@ TEST_P(ParameterizedFp8GemmRewriteTest,
       )");
 }
 
+TEST_P(ParameterizedFp8GemmRewriteTest, Rank3ScaledABUnscaledDVectorBiasF8) {
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/tests/gemm_rewrite_test.cc:5326]: Comment this tests a case where there is a slice that does not chop off the ends of dimensions.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 25]
```diff
@@ -112,27 +112,36 @@ Shape PadShapeToMultipleOf16(const Shape old_shape,
...
 
-// Pad the non-batch dimensions of the operands to multiples of 16 as required
-// by cuBLASLt.
-HloInstruction *PadOperandToMultipleOf16(absl::Span<const int64_t> batch_dims,
-                                         HloInstruction *instr,
-                                         HloInstruction *x) {
+// Pad the non-batch dimensions of the operands to the target shape.
+HloInstruction *PadOperandToTargetShape(const Shape &target,
+                                        HloInstruction *instr,
+                                        HloInstruction *x) {
```
[the remark this answers is not in the document]
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:118]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 22 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1134]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 23 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1171]: Resolved.

File: tensorflow/compiler/xla/service/gpu/tests/gemm_rewrite_test.cc (hunk 24 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/tests/gemm_rewrite_test.cc:5326]: Resolved.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 26]
```diff
@@ -1103,7 +1131,22 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
+    // if (slice && bitcast) {
+    if (bitcast) {
+      maybe_constant_folded_bias =
+          instr->AddInstruction(HloInstruction::CreateBitcast(
+              slice->shape(), maybe_constant_folded_bias));
+    }
+
+    if (gemm->custom_call_target() == kCublasLtMatmulF8CallTarget ||
+        gemm->custom_call_target() == kCublasLtMatmulCallTarget ||
+        gemm->custom_call_target() == kGemmCallTarget) {
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1144]: I think `gemm` is guaranteed to be a cublas or cublas LT custom call, so no need to check this.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 26 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:1144]: Read from [here](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/gpu/cublas_cudnn.cc;l=29?q=IsLegacyCublasMatmul&sq=&ss=tensorflow%2Ftensorflow) there are 3 kinds of cublas or cublaslt calls.

**Review by reedwm — APPROVED** [review]

**Comment by reedwm** [comment]
@wenscarl thanks for addressing all my comments! Can you resolve conflicts?

**Comment by wenscarl** [comment]
> @wenscarl thanks for addressing all my comments! Can you resolve conflicts?

The gemm_rewrite_test did not pass for the current master. Could you take a look?

**Comment by reedwm** [comment]
> The gemm_rewrite_test did not pass for the current master. Could you take a look?

It's passing for me. I tried on https://github.com/openxla/xla/commit/d7b3b230abca3bc7adafeff7344aaafc6e6994dc with `bazel test //xla/service/gpu/tests:gemm_rewrite_test` and it passes. What commit are you synced to and what command are you using to run?

**Comment by reedwm** [comment]
I also tried on the TF repo at 35af49017480ee112fc8c9850862f1eb61d611ee, since this PR is on the TF repo. It still passes for me.

**Comment by wenscarl** [comment]
> I also tried on the TF repo at [35af490](https://github.com/tensorflow/tensorflow/commit/35af49017480ee112fc8c9850862f1eb61d611ee), since this PR is on the TF repo. It still passes for me.

I tried the same commit, Used `TF_CUDA_CLANG=0` in .bazelrc. BatchRowTransposeFoldCheck, BatchFromMinorDimTransposeIsNotFolded, BatchedInstrLayoutTransposed, BatchedInstrLayoutBatchNotInMinorDim failed.

**Comment by reedwm** [comment]
I'm also using `TF_CUDA_CLANG=0`. I'm using CUDA 12.1.1 with cuDNN 8.9.1. What version of CUDA/cuDNN are you using? And what's the error message?

**Comment by wenscarl** [comment]
> I'm also using `TF_CUDA_CLANG=0`. I'm using CUDA 12.1.1 with cuDNN 8.9.1. What version of CUDA/cuDNN are you using? And what's the error message?
Seeing :
```
Mismatch count 4 (10.0000%) in shape f32[2,5,4] (40 elements), abs bound 1e-05, rel bound 1e-05
Top relative error mismatches:
  actual         0.000914607197, expected         0.000930683687, index {1,2,2}, rel error   0.0173, abs error 1.61e-05
```
while testing BatchedInstrLayoutTransposed on Hopper.

**Comment by reedwm** [comment]
What version of CUDA and cuDNN are you using?

**Comment by wenscarl** [comment]
> What version of CUDA and cuDNN are you using?

cuda 12.1 + cudnn 8.9.2

**Comment by wenscarl** [comment]
> I think it's fine to just increase the test tolerance, since the error is almost within tolerances already.

Fixes by #61415

**Comment by reedwm** [comment]
@wenscarl can you sync past #61415?

**Review by reedwm — APPROVED** [review]
